tpm.go

```diff
@@ -0,0 +1,928 @@
+			} else {
+				err = xerrors.Errorf("download from %s failed: %w", issuerUrl, e)
+			}
+		}
+
+		if err != nil {
+			return nil, xerrors.Errorf("cannot download parent certificate of %v: %w", cert.Subject, err)
+		}
+
+		out = append(out, parent.Raw)
+		cert = parent
+	}
+
+	return out, nil
+}
+
+// fetchEkCertificateChain will attempt to obtain the entire EK certificate chain, first by reading the EK certificate from
+// the TPM and then fetching all of the parent certificates. If parentsOnly is true, the returned data will not include the
+// actual EK certificate.
+func fetchEkCertificateChain(tpm *tpm2.TPMContext, parentsOnly bool) (*ekCertData, error) {
+	var data ekCertData
+
+	if cert, err := readEkCertFromTPM(tpm); err != nil {
+		return nil, xerrors.Errorf("cannot obtain endorsement key certificate from TPM: %w", err)
+	} else {
+		if !parentsOnly {
+			data.Cert = cert
+		}
+
+		c, err := x509.ParseCertificate(cert)
+		if err != nil {
+			return nil, xerrors.Errorf("cannot parse endorsement key certificate: %w", err)
+		}
+
+		parents, err := fetchParentCertificates(c)
+		if err != nil {
+			return nil, xerrors.Errorf("cannot obtain parent certificates for %s: %w", c.Subject, err)
+		}
+		data.Parents = parents
+	}
+
+	return &data, nil
+}
+
+// saveEkCertificateChain will save the supplied EK certificate chain to the file at the specified path. The file will be updated
+// atomically.
+func saveEkCertificateChain(data *ekCertData, dest string) error {
+	f, err := osutil.NewAtomicFile(dest, 0600, 0, sys.UserID(osutil.NoChown), sys.GroupID(osutil.NoChown))
+	if err != nil {
+		return xerrors.Errorf("cannot create new atomic file: %w", err)
+	}
+	defer f.Cancel()
+
+	if err := tpm2.MarshalToWriter(f, data); err != nil {
+		return xerrors.Errorf("cannot marshal cert chain: %w", err)
+	}
+
+	if err := f.Commit(); err != nil {
+		return xerrors.Errorf("cannot atomically replace file: %w", err)
+	}
+
+	return nil
+}
+
+// FetchAndSaveEKCertificateChain attempts to obtain the endorsement key certificate for the TPM associated with the tpm parameter,
+// download the parent certificates and then save them atomically to the specified file in a form that can be decoded by
+// SecureConnectToDefaultTPM. This function requires network access.
+//
+// This depends on the presence of the Authority Information Access extension in the leaf certificate and any intermediate certificates,
+// which must contain a URL for the parent certificate.
+//
+// This function will stop when it encounters a certificate that doesn't specify an issuer URL, or when it encounters a self-signed
+// certificate.
+//
+// If no endorsement key certificate can be obtained, an error will be returned.
+//
+// If parentsOnly is true, this function will only save the parent certificates as long as the endorsement key certificate can be
+// reliably obtained from the TPM.
+func FetchAndSaveEKCertificateChain(tpm *TPMConnection, parentsOnly bool, destPath string) error {
```

Comment:
why fetching and saving together is more convenient it doesn't feel very orthogonal also it forces to save to a file. It seems a form that is fetch only and returns (ekCert *x509.Certificate, parents []*x509.Certificate) would be more general and can be combined with either Encode or Save as useful.  A form that only does the readEkCertFromTPM might also be useful for flexibility?

Reply:
The reason this is here is because saving them to a file will be the most common way of storing them, but it's probably worth having a FetchEKCertificateChain which returns (*x509.Certificate, []*x509.Certificate, error) as well. I only really added EncodeEKCertificateChain originally because it allows the tests to use SecureConnectToDefaultTPM without having to save a file.

Reply:
yes, let's reconsider adding that in a follow up